staging/src/k8s.io/apiserver/pkg/endpoints/filters/actas.go

```diff
@@ -0,0 +1,245 @@
+			klog.V(4).Infof("%v", err)
+			responsewriters.InternalError(w, req, err)
+			return
+		}
+		if len(actAsRequests) == 0 {
+			handler.ServeHTTP(w, req)
+			return
+		}
+
+		ctx := req.Context()
+		requestor, exists := request.UserFrom(ctx)
+		if !exists {
+			responsewriters.InternalError(w, req, errors.New("no user found for request"))
+			return
+		}
+
+		attributes, err := GetAuthorizerAttributes(ctx)
+		if err != nil {
+			responsewriters.InternalError(w, req, err)
+			return
+		}
+
+		// if groups are not specified, then we need to look them up differently depending on the type of user
+		// if they are specified, then they are the authority (including the inclusion of system:authenticated/system:unauthenticated groups)
+		groupsSpecified := len(req.Header[ActAsGroupHeader]) > 0
+
+		// make sure we're allowed to actas each thing we're requesting.  While we're iterating through, start building username
+		// and group information
+		username := ""
+		groups := []string{}
+		userExtra := map[string][]string{}
+		uid := ""
+		for _, actAsRequest := range actAsRequests {
+			gvk := actAsRequest.GetObjectKind().GroupVersionKind()
+			actingAsAttributes := &authorizer.AttributesRecord{
+				User:            requestor,
+				Verb:            "actas",
+				APIGroup:        gvk.Group,
+				APIVersion:      gvk.Version,
+				Namespace:       actAsRequest.Namespace,
+				Name:            actAsRequest.Name,
+				ResourceRequest: true,
+			}
+
+			switch gvk.GroupKind() {
+			case v1.SchemeGroupVersion.WithKind("ServiceAccount").GroupKind():
+				actingAsAttributes.Resource = "serviceaccounts"
+				username = serviceaccount.MakeUsername(actAsRequest.Namespace, actAsRequest.Name)
+				if !groupsSpecified {
+					// if groups aren't specified for a service account, we know the groups because its a fixed mapping.  Add them
+					groups = serviceaccount.MakeGroupNames(actAsRequest.Namespace)
+				}
+
+			case v1.SchemeGroupVersion.WithKind("User").GroupKind():
+				actingAsAttributes.Resource = "users"
+				username = actAsRequest.Name
+
+			case v1.SchemeGroupVersion.WithKind("Group").GroupKind():
+				actingAsAttributes.Resource = "groups"
+				groups = append(groups, actAsRequest.Name)
+
+			case authenticationv1.SchemeGroupVersion.WithKind("UserExtra").GroupKind():
+				extraKey := actAsRequest.FieldPath
+				extraValue := actAsRequest.Name
+				actingAsAttributes.Resource = "userextras"
+				actingAsAttributes.Subresource = extraKey
+				userExtra[extraKey] = append(userExtra[extraKey], extraValue)
+
+			case authenticationv1.SchemeGroupVersion.WithKind("UID").GroupKind():
+				uid = string(actAsRequest.Name)
+				actingAsAttributes.Resource = "uids"
+
+			default:
+				klog.V(4).InfoS("unknown actas request type", "request", actAsRequest)
+				responsewriters.Forbidden(ctx, actingAsAttributes, w, req, fmt.Sprintf("unknown actas request type: %v", actAsRequest), s)
+				return
+			}
+
+			decision, reason, err := a.Authorize(ctx, actingAsAttributes)
```

Comment:
let's start by modelling this as, "does the requesting user have permission to ActAs at all. If so, allow any ActAs".